src/datastructure/RailwayNetwork.cpp

```diff
@@ -653,3 +657,749 @@ cda_rail::Network::Network(const std::filesystem::path &p) {
+            throw std::runtime_error("Something went wrong, vertex has more than one neighbor still.");
+        }
+
+        // Process edge pair
+        const auto edge_pair_index = *vertex_neighbors[j].begin();
+        const auto& edge_pair = edge_pairs[edge_pair_index];
+
+        const auto& edge = get_edge(edge_pair.first);
+        vertex_neighbors[edge.source].erase(edge_pair_index);
+        vertex_neighbors[edge.target].erase(edge_pair_index);
+
+        // Get next vertex
+        if (edge.source != j && edge.target == j) {
+            ret_val.emplace_back(edge_pair.second, edge_pair.first);
+            j = edge.source;
+        }
+        else if (edge.target != j && edge.source == j) {
+            ret_val.emplace_back(edge_pair.first, edge_pair.second);
+            j = edge.target;
+        } else {
+            throw std::runtime_error("Something went wrong, source and target are not as expected.");
+        }
+    }
+
+    for (int i = 0; i < number_of_vertices(); ++i) {
+        if (vertex_neighbors[i].size() > 0) {
+            throw std::runtime_error("Something went wrong, not everything was processed.");
+        }
+    }
+
+    return ret_val;
+
+}
+
+std::vector<int> cda_rail::Network::inverse_edges(const std::vector<int> &edge_indices,
+                                                  const std::vector<int> &edges_to_consider) const {
+    /**
+     * Returns a vector of edge indices that are the inverse of the given edge indices, i.e.,
+     * return edges_to_consider - edge_indices
+     *
+     * @param edge_indices: Vector of edge indices
+     * @param edges_to_consider: Vector of edge indices to consider
+     * @return: Vector of edge indices that are the inverse of the given edge indices
+     */
+
+    // Initialize return value
+    std::vector<int> ret_val;
+
+    // Check if all edges exist
+    for (const auto& edge_index: edge_indices) {
+        if (!has_edge(edge_index)) {
+            throw std::invalid_argument("Edge does not exist");
+        }
+    }
+    for (const auto& edge_index: edges_to_consider) {
+        if (!has_edge(edge_index)) {
+            throw std::invalid_argument("Edge does not exist");
+        }
+    }
+
+    // ret_val = edges_to_consider - edge_indices
+    for (const auto& edge_index: edges_to_consider) {
+        if (std::find(edge_indices.begin(), edge_indices.end(), edge_index) == edge_indices.end()) {
+            ret_val.emplace_back(edge_index);
+        }
+    }
+
+    return ret_val;
+}
+
+int cda_rail::Network::max_vss_on_edge(int index) const {
+    /**
+     * Returns how many vss can be placed on a certain edge at most.
+     */
+
+    const auto& edge = get_edge(index);
+    if (!edge.breakable || edge.min_block_length == 0) {
+        return 0;
+    }
+    return static_cast<int>(std::floor(edge.length / edge.min_block_length));
+}
+
+cda_rail::MultiArray<double> cda_rail::Network::all_edge_pairs_shortest_paths() const {
```

Comment:
Why the need for `MultiArray`? Isn't a vector of vectors sufficient?

Reply:
True, no need. Done